src/llmescache/langchain.py

```diff
@@ -0,0 +1,124 @@
+            es_index (str): The name of the index to use for the cache store.
+            store_input (bool): Whether to store the LLM input in the cache, i.e., the input prompt.
+                Defaults to True.
+            store_timestamp (bool): Whether to store the datetime in the cache, i.e., the time of the
+                first request for an input. Defaults to True.
+            store_input_params (bool): Whether to store the input parameters in the cache, i.e., the
+                parameters used to generate the LLM input. Defaults to True.
+            metadata (Optional[dict], optional): Additional metadata to store in the cache, for filtering purposes.
+                This must be JSON serializable. Defaults to None.
+        """
+
+        self._es_client = es_client
+        self.index = es_index
+
+        self.store_input = store_input
+        self.store_timestamp = store_timestamp
+        self.store_input_params = store_input_params
+        self.metadata = metadata or {}
+
+        if not self._es_client.ping():
+            raise elasticsearch.exceptions.ConnectionError(
+                "ElasticSearch cluster is not available, not able to set up the cache store."
+            )
+
+        if not self._es_client.indices.exists(index=self.index):
+            self._es_client.indices.create(
+                index=self.index, body=self.get_default_mapping
+            )
+
+    @property
+    def get_default_mapping(self) -> Dict[str, Any]:
+        """Get the default mapping for the index."""
+        return {
+            "mappings": {
+                "properties": {
+                    "llm_output": {"type": "text"},
+                    "llm_params": {"type": "text"},
+                    "llm_input": {"type": "text"},
+                    "metadata": {"type": "object"},
+                    "date": {"type": "date"},
+                }
+            }
+        }
+
+    @staticmethod
+    def _key(prompt: str, llm_string: str) -> str:
+        """Generate a key for the cache store."""
+
+        def _hash(_input: str) -> str:
+            return hashlib.md5(_input.encode()).hexdigest()
+
+        return _hash(prompt + llm_string)
+
+    def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
+        """Look up based on prompt and llm_string."""
+        try:
+            record = self._es_client.get(
+                index=self.index, id=self._key(prompt, llm_string)
+            )
+            return _loads_generations(record["_source"]["llm_output"])
+        except elasticsearch.exceptions.NotFoundError:
+            return None
+
+    def update(self, prompt: str, llm_string: str, return_val: RETURN_VAL_TYPE) -> None:
+        """Update based on prompt and llm_string."""
+        body = {
+            "llm_output": _dumps_generations(return_val),
+        }
+
+        if self.store_input_params:
+            body["llm_params"] = llm_string
+
+        if self.metadata:
```

Comment:
I would not set metadata as an empty dict, bu verify here that is not None, so we can remove `type ignore` , that we should always avoid in prod

Reply:
trying to figure out how to let mypy be ok